### DISCUSSION TITLE
Add a MongoDB Motor Repository or Plugin

### DISCUSSION THREAD

**Discussion by Goldziher** [litestar-org/litestar#1559] [Ideas]
I'd suggest we add a MongoDB repository to work directly with the Motor async driver. The issue with the driver is that it has no official types, just an unofficial (but very solid) stubs library. 

There are of course multiple ODMs such as beanie, Odmantic, Mongox that build on Motor + pydantic to create a smoother experience. I do think though that these have some inherent problems -

1. None of them uses the proper mongo query and aggregation syntax. They try to make it more `pythonic`, but i don't personally like that. 

2. They are not performant or performance orientated. 

3. They don't integrate as smoothly with us as a proper repository or plugin will. 

Thoughts and suggestions?

**Comment by chris-telemetry** [comment]
We're using MongoDB w/ Litestar at Telemetry.

We use pure Motor with no ODM.

Addressing your points in order:
1. I agree. Like Entity Framework, it certainly fills a need - but eventually you end up needing to use the proper syntax and you end up with two different patterns. That's not terrible but I personally don't enjoy it.
2. Yeah this was why we ended up using pure Motor. At least for us, performance is already a major issue. Mongo emphasizes nesting and Pydantic especially struggles with nested models.
3. Agreed. I'd also say there's a pretty big risk with tying yourself to an ODM. In my opinion, the Mongo-Python ecosystem is pretty poor and unstable. Motor itself was only recently officially adopted by MongoDB and Motor itself is just a ThreadPoolExecutor wrapper around PyMongo.


That all being said, the popularity of Mongo combined with the deficiencies in the ecosystem could be an opportunity for Litestar to fill a need. I'd of course be willing to assist.

**Reply by Goldziher** [reply]
> We're using MongoDB w/ Litestar at Telemetry.
> 
> We use pure Motor with no ODM.
> 
> Addressing your points in order:
> 1. I agree. Like Entity Framework, it certainly fills a need - but eventually you end up needing to use the proper syntax and you end up with two different patterns. That's not terrible but I personally don't enjoy it.
> 2. Yeah this was why we ended up using pure Motor. At least for us, performance is already a major issue. Mongo emphasizes nesting and Pydantic especially struggles with nested models.
> 3. Agreed. I'd also say there's a pretty big risk with tying yourself to an ODM. In my opinion, the Mongo-Python ecosystem is pretty poor and unstable. Motor itself was only recently officially adopted by MongoDB and Motor itself is just a ThreadPoolExecutor wrapper around PyMongo.
> 
> 
> That all being said, the popularity of Mongo combined with the deficiencies in the ecosystem could be an opportunity for Litestar to fill a need. I'd of course be willing to assist.

Great, wanna submit a PR?

**Reply by 0asys** [reply]
Would be cool :)
